src/oc/hocdec.h

```diff
@@ -140,17 +143,85 @@ struct Symbol { /* symbol table entry */
 
 using hoc_List = hoc_Item;
 
-union Datum { /* interpreter stack type */
-    double val;
-    Symbol* sym;
-    int i;
-    double* pval; /* first used with Eion in NEURON */
-    Object** pobj;
-    Object* obj; /* sections keep this to construct a name */
-    char** pstr;
-    hoc_Item* itm;
-    hoc_List* lst;
-    void* _pvoid; /* not used on stack, see nrnoc/point.cpp */
+namespace neuron::detail {
+struct _Datum { /* interpreter stack type */
+    union {
+        double val;
+        Symbol* sym;
+        int i;
+        double* pval; /* first used with Eion in NEURON */
+        Object** pobj;
+        Object* obj; /* sections keep this to construct a name */
+        char** pstr;
+        hoc_Item* itm;
+        hoc_List* lst;
+        void* _pvoid; /* not used on stack, see nrnoc/point.cpp */
+    };
+};
+}  // namespace neuron::detail
+
+struct Datum: private neuron::detail::_Datum {
+    template <typename T>
+    [[nodiscard]] T& literal_value() {
+        static_assert(std::is_standard_layout_v<T>);
+        return *reinterpret_cast<T*>(&*this);  // Eww
+    }
+
+    template <typename T>
+    explicit operator T() const {
+        if constexpr (std::is_same_v<T, double>) {
+            return val;
+        } else if constexpr (std::is_same_v<T, Symbol*>) {
+            return sym;
+        } else if constexpr (std::is_same_v<T, int>) {
+            return i;
+        } else if constexpr (std::is_same_v<T, double*>) {
+            return pval;
+        } else if constexpr (std::is_same_v<T, Object**>) {
+            return pobj;
+        } else if constexpr (std::is_same_v<T, Object*>) {
+            return obj;
+        } else if constexpr (std::is_same_v<T, char**>) {
+            return pstr;
+        } else if constexpr (std::is_same_v<T, hoc_Item*>) {
+            return itm;
+        } else if constexpr (std::is_same_v<T, hoc_List*>) {
+            return lst;
+        } else if constexpr (std::is_pointer_v<T> || std::is_same_v<T, std::nullptr_t>) {
+            return static_cast<T>(_pvoid);
+        } else {
+            static_assert(!std::is_same_v<T, T>, "Type not supported");
+        }
+        return static_cast<T>(*this);
```

Comment:
Maybe `return literal_value<T>()` is sufficient as an implementation of this operator?

Reply:
For sure, just wanted to catch all usages

Reply:
Will have to drop const in the function signature though